include/net/ethernet.h

```diff
@@ -376,6 +376,51 @@ struct net_eth_hdr *net_eth_fill_header(struct ethernet_context *ctx,
 					u8_t *src,
 					u8_t *dst);
 
+#if defined(CONFIG_NET_L2_ETHERNET_MGMT)
+void ethernet_mgmt_raise_carrier_on_event(struct net_if *iface);
+void ethernet_mgmt_raise_carrier_off_event(struct net_if *iface);
+
+/**
+ * @brief Inform ethernet L2 driver that ethernet carrier is detected.
+ * This happens when cable is connected.
+ *
+ * @details Note that as this function is typically called by network
+ * device driver in ISR context, we just raise an event here and let the
+ * ethernet L2 driver catch that and act as needed.
+ *
+ * @param iface Network interface
+ */
+static inline void net_eth_carrier_on(struct net_if *iface)
+{
+	ethernet_mgmt_raise_carrier_on_event(iface);
```

Comment:
hum that's not what I meant in my previous comment.

net_eth_carrier_on/off would be publicly declared here the same signature you did, but implemented in ethernet l2. There you would directly do what's necessary for putting if up, if carrier down etc... and raising these net mgmt event. The idea is that device drivers should not raise net mgmt event themselves (besides very specific uses case). And doing so, you could built ethernet drivers/l2 without the need of net mgmt (if reducing the memory footprint is necessary). Note that this will also reduce the code complexity in ethernet l2 (as it won't need to listen to net mgmt)

Reply:
> There you would directly do what's necessary for putting if up, if carrier down etc... and raising these net mgmt event.

I was thinking that but decided to implement it like this. The reason being that I would not like to call various net_if functions from ISR context.

Reply:
Still, as a comparision in linux it would mean relying on netlink into drivers to pass message to net core. You don't want that.

Most, if not all, ethernet drivers have their own thread. Using it would just be fine (in case mcux there is a eth_mcux_phy_event() for instance)

Reply:
> Most, if not all, ethernet drivers have their own thread. Using it would just be fine (in case mcux there is a eth_mcux_phy_event() for instance)

It is impossible to say in what context the driver would call IP stack and how much stack there is for anything. In this respect the only safe way to pass this information is via mgmt event. Other option would be to use work queue for this but it looks a bit cumbersome to use.
IMHO the event passing from driver to upper stack looks quite good and convenient, I am not sure why you are so concerned about it.

Reply:
I thought bringing the linux/netlink comparison would tell the reason. It's not "reliable". It's true that it is convenient, but logically we should not tie kind of critical system event to a user messaging event subsystem.

Reply:
We can certainly use work queue here if needed but I am not buying the "reliable" thingy. The carrier ON/OFF is quite similar as network interface ON/OFF event, and if we loose the iface ON/OFF event then that is bad too.
This is a simple issue, and there is no point of arguing this. I will prepare a version that uses the work queue instead.

Reply:
Sure I don't think there is any net core part using on/off event though. dhcpv4 is more like a user thing somehow